### DISCUSSION TITLE
Flatten JSon

### DISCUSSION THREAD

**Discussion by NashTools** [stephenberry/glaze#1067] [Q&A]
I would like to efficiently parse similar JSon strings from the Binance Api into the same struct. It seems that I can map multiple names from JSon strings to the same member variable with meta objects like:
`static constexpr auto value = object("p", glz::quoted_num<&BinanceTradeGlaze::p>, "price", glz::quoted_num<&BinanceTradeGlaze::p> ...
`
Is this correct?

However, even so I have a problem, because in one type of JSon the relevant price field is in an inner object of the Json like this:
`{"e":"ORDER_TRADE_UPDATE","T":1684156948198,"E":1684156948203,"o":{"p":"27363.90" ...
`
Is there a way to map `o:p` in the JSon to the member variable `p` in my struct? I added `o:p` to the meta object, but this doesn't seem to work.
I understand that I could just define an inner struct and it would parse it correctly, but I would like to only have one type of struct for a simpler interface to the methods that use the Binance trades and also do as little parsing and copying as possible.

**Comment by stephenberry** [comment]
It is best to define the inner struct with Glaze because it is telling the compiler the actual layout of the JSON, which makes parsing more efficient. However, you can use pointers and references in C++ to remap this inner value to a value that is higher up in your struct to make your interface simpler.

```c++
struct inner
{
   double p{};
};
```

```c++
struct outer
{
   std::string e{};
   uint64_t T{};
   uint64_t E{};
   inner o{};
   double* p{&o.p}; // a reference or pointer here remaps the value, but a reference cannot use pure reflection and will need a glz::meta
};
```

Is this the kind of remapping you want? If you could provide the C++ struct you want, that would help with explaining a solution.

**Reply by NashTools** [reply]
Thanks, these are 2 different JSons that I wanted to parse into the same struct, I'm currently only interested in parsing JSons, not writing them:
```
{"e":"trade","E":1666781982803,"T":1666781982799,"s":"BTCBUSD","t":321337415,"p":"20659.8","q":"0.044","X":"MARKET","m":true}

{"e":"ORDER_TRADE_UPDATE","T":1684156948198,"E":1684156948203,"o":{"s":"BTCUSDT","c":"1764","S":"BUY","o":"LIMIT","f":"GTX","q":"0.005","p":"27363.90","ap":"27363.90000","sp":"0","x":"TRADE","X":"FILLED","i":150430598779,"l":"0.001","z":"0.001","L":"27363.90","n":"0.00547278","N":"USDT","T":1684156948198,"t":3715937249,"b":"0","a":"27.36622","m":true,"R":false,"wt":"CONTRACT_PRICE","ot":"LIMIT","ps":"BOTH","cp":false,"rp":"0","pP":false,"si":0,"ss":0}}
```
This is the struct that I was using so far:
```
struct BinanceTrade {
    double price;
    double amount;
    double buyAmount;
    double sellAmount;
    int64_t eventTime;
    int64_t tradeTime;
    int64_t identifier;
    bool sell;
    bool liquidation;
};
```
And this is a method that uses the struct:
```
void AggTradeList::addTrade(const BinanceTrade &trade) {
    int64_t tradeTime = trade.tradeTime;
    int64_t tradeId = trade.identifier;
    int64_t tradeIndex = tradeTime % milliSeconds;
   ...
```
I think there probably is no good way to parse both strings into the same struct. Before I was parsing them manually with a simple string search, which was already faster than most other JSon libraries, but glaze is even faster than my manual approach. I just tried implementing your suggestion with creating 2 more structs
```
struct BinanceTradePrivateInner { ...

struct BinanceTradePrivate {
   ...
   BinanceTradeInner o;
   double& p{o.p}
   ...
```
but the problem with this approach is that the references still take up memory. I was hoping the compiler would basically optimize them away, but they do increase the sizeof of the BinanceTradePrivate struct.
I'm thinking about instead of creating references to the inner members, to just have the same getters in BinanceTrade and BinanceTradePrivate and make AggTradeList::addTrade a template and just use the getters. This would still be a relatively clean interface. Unless of course there is a better way of doing this.

**Reply by stephenberry** [reply]
Your idea of getters seems like a good approach. This is a bit tricky because you're trying to parse two different JSON layouts into the same struct. However, it does look like your second layout doesn't contain `"p"` in the top level object. This means that we can avoid the reference (and the increase in size) by using a lambda in `glz::meta` to remap the value.

Here is a rework of my original example using a lambda to avoid the size increase:
```c++
struct inner
{
   double p{};
};

struct outer
{
   std::string e{};
   uint64_t T{};
   uint64_t E{};
   inner o{};
};

template <>
struct glz::meta<outer>
{
   using V = outer;
   static constexpr auto value = object(&V::e, &V::T, &V::E, "p", [](auto& self) -> auto& { return self.o.p; });
};
```

In this way we still use the inner struct to express in C++ our layout, but we allow the `"p"` field to be parsed from the top level. In this way it can handle parsing either the nested `"o"` field, or it can parse `"p"` into the same C++ structure. And, the lambda adds no size overhead to your struct. This approach should give you the ability to parse either format and not increase memory size or lose performance.
